### ISSUE TITLE
NVIDIA Driver installation fails (version 11.3) - Azure

### ISSUE THREAD

**Issue by brentwg** [ComputeCanada/puppet-magic_castle#138]
I have repeatedly attempted to launch the following instance type as a GPU compute node in Azure using version 11.3 --> Standard NC4as_T4_v3 (4 vcpus, 28 GiB memory); however, the NVIDIA driver installation fails and the node is listed as drained. I know that it can sometimes take up to ~45 minutes for all of the puppet config to run successfully. I've waited more than 24 hours and the nodes still remain in the same state.

I can see this in the logs:

`Execution of '/usr/bin/yum -d 0 -e 0 -y install nvidia-driver-latest-dkms' returned 1: Error: Package: 3:nvidia-driver-latest-dkms-cuda-470.57.02-1.el7.x86_64 (cuda)
        Requires: opencl-filesystem
Error: Package: 3:kmod-nvidia-latest-dkms-470.57.02-1.el7.x86_64 (cuda)
        Requires: dkms
Error: Package: 3:nvidia-driver-latest-dkms-cuda-470.57.02-1.el7.x86_64 (cuda)
        Requires: ocl-icd
You could try using --skip-broken to work around the problem
You could try running: rpm -Va --nofiles --nodigest
(/Stage[main]/Profile::Gpu::Install::Passthrough/Package[nvidia-driver-latest-dkms]/ensure) change from 'purged' to 'present' failed: Execution of '/usr/bin/yum -d 0 -e 0 -y install nvidia-driver-latest-dkms' returned 1: Error: Package: 3:nvidia-driver-latest-dkms-cuda-470.57.02-1.el7.x86_64 (cuda)
(/Stage[main]/Profile::Gpu::Install::Passthrough/Package[nvidia-driver-latest-dkms]/ensure)            Requires: opencl-filesystem
(/Stage[main]/Profile::Gpu::Install::Passthrough/Package[nvidia-driver-latest-dkms]/ensure) Error: Package: 3:kmod-nvidia-latest-dkms-470.57.02-1.el7.x86_64 (cuda)
(/Stage[main]/Profile::Gpu::Install::Passthrough/Package[nvidia-driver-latest-dkms]/ensure)            Requires: dkms
(/Stage[main]/Profile::Gpu::Install::Passthrough/Package[nvidia-driver-latest-dkms]/ensure) Error: Package: 3:nvidia-driver-latest-dkms-cuda-470.57.02-1.el7.x86_64 (cuda)
(/Stage[main]/Profile::Gpu::Install::Passthrough/Package[nvidia-driver-latest-dkms]/ensure)            Requires: ocl-icd
(/Stage[main]/Profile::Gpu::Install::Passthrough/Package[nvidia-driver-latest-dkms]/ensure)  You could try using --skip-broken to work around the problem
(/Stage[main]/Profile::Gpu::Install::Passthrough/Package[nvidia-driver-latest-dkms]/ensure)  You could try running: rpm -Va --nofiles --nodigest`

I tried the same experiment with 11.2 and, although I see the same error, eventually the dependencies get installed and everything is fine. Then SLURM lists the node as idle. If there is any additional information I can pass along, please let me know. I am happy to help resolve this or discover that there is maybe a problem on my end. In the meantime, I am thinking about rolling back to 11.2. Please advise.

**Comment by cmd-ntrf** [comment]
I have fixed the epel dependency issue in puppet-magic_castle main branch.

To try it replace `config_version` value in your `main.tf` by `main`.
See: https://github.com/ComputeCanada/magic_castle/blob/main/examples/azure/main.tf#L8

This should fix the errors you included in your issue. We should then be able to find out following errors that could explain why the node does not show up as idle in Slurm.

**Comment by brentwg** [comment]
Thanks very much for looking into this. To answer your first question, the image I am using is: 

image = {
    publisher = "OpenLogic",
    offer     = "CentOS-CI",
    sku       = "7-CI"
  }

I replaced the `config_version` with `main` as suggested and ran the code again. During the first iteration of the puppet code, the NVIDIA driver installation failed again. However, I manually executed `puppet agent --test` and now the NVIDIA is installed and SLURM is reporting the node as idle. Is that how it's supposed to work? I mean, does the driver actually get installed during the second iteration? Or is something still broken? 

Again, if you require any additional information, I am happy to help. Thanks again for responding so quickly.

**Comment by cmd-ntrf** [comment]
Could you dump the entire puppet log for the compute node : `journalctl -u puppet > puppet_dump.log` and create a [gist](https://gist.github.com/) with it? I will take a look to see what we are missing.

**Comment by brentwg** [comment]
`puppet_dump.log` contents have been saved in this [gist](https://gist.github.com/brentwg/97817c1578fc7efb1e0335f593e43494).

**Comment by cmd-ntrf** [comment]
Based on the log provided, `epel` yum repository was created after puppet tried to install gpu rpms. It should have been fixed in the main branch.

Was this cluster from scratch ?

**Comment by brentwg** [comment]
Not a cluster from scratch. It was originally created a few weeks ago.

**Comment by cmd-ntrf** [comment]
Applying the fix on a live cluster is a bit different.
From the login node as user centos
- `ssh puppet`
- `cd /etc/puppetlabs/code/environment/production`
- `sudo git checkout main`
- `sudo git pull`

Then rebuild your compute node or add a new node.

**Comment by brentwg** [comment]
I get it. Thank you.

**Comment by brentwg** [comment]
Quick question. When I navigate to the puppet directory listed above, and I run `git status` I get the following:

`# HEAD detached at 11.3
# Untracked files:
#   (use "git add <file>..." to include in what will be committed)
#
#	.librarian/
#	.tmp/
#	Puppetfile.lock
#	data/user_data.yaml
#	modules/
#	site/profile/facts.d/terraform_facts.yaml
nothing added to commit but untracked files present (use "git add" to track)`

Is this a problem? Do I need to commit anything first? (`terraform_facts.yaml`, I am looking at you). Or can I just clobber everything? Please advise.

**Comment by cmd-ntrf** [comment]
Not a problem. These files are necessary, but not part of the repo.
I am not sure what you mean by _clobber_, but do not delete them.

If you checkout the main branch and pull the changes from origin, the files will be left as they are.
